daemon/container_operations.go

```diff
@@ -940,11 +940,12 @@ func (daemon *Daemon) disconnectFromNetwork(container *container.Container, n li
 
 func (daemon *Daemon) tryDetachContainerFromClusterNetwork(network libnetwork.Network, container *container.Container) {
 	if daemon.clusterProvider != nil && network.Info().Dynamic() && !container.Managed {
-		if err := daemon.clusterProvider.DetachNetwork(network.Name(), container.ID); err != nil {
-			logrus.Warnf("error detaching from network %s: %v", network.Name(), err)
-			if err := daemon.clusterProvider.DetachNetwork(network.ID(), container.ID); err != nil {
-				logrus.Warnf("error detaching from network %s: %v", network.ID(), err)
-			}
+		if err := daemon.clusterProvider.DetachNetwork(network.ID(), container.ID); err != nil {
+			logrus.Warnf("error detaching from network %s: %v", network.ID(), err)
+		}
+		// We need delete the network before containerStop returns, otherwise 'docker restart' would get error.
+		if err := daemon.DeleteManagedNetwork(network.ID()); err != nil {
+			logrus.WithError(err).Warnf("error deleteing network %s", network.ID())
 		}
```

Comment:
If we failed to detach the container from the network, should we still try to delete the network? I think this would always fail in that case (as there's still endpoints attached)? 🤔

Reply:
Yes, if `DetachNetwork` fails, then there's no point continuing to delete the network. I will modify it.

There is only one possible error in` DetachNetwork`: `could not find network attachment ` in `c.attachers[aKey]`, I don't know how this could happen.

Reply:
Modified per your suggestion.

Reply:
After thinking, I still feel we should run `daemon.DeleteManagedNetwork` no matter `DetachNetwork` fails or not, because this is nearly the final stage of container-stop cleanup and network deletion is not run elsewhere when container runs with attachable overlay network.